pcp/PCPMachine.c

```diff
@@ -50,40 +50,97 @@ static void PCPMachine_updateCPUcount(PCPMachine* this) {
+      sysname.cp = NULL;
+   else if (String_eq(sysname.cp, "Linux"))
+      this->sys = SYSTEM_NAME_LINUX;
+   else if (String_eq(sysname.cp, "Darwin"))
+      this->sys = SYSTEM_NAME_DARWIN;
+   free(sysname.cp);
+}
+
+static void PCPMachine_updateLinuxMemoryInfo(PCPMachine* this) {
+   Machine* super = &this->super;
    unsigned long long int freeMem = 0;
    unsigned long long int swapFreeMem = 0;
    unsigned long long int sreclaimableMem = 0;
-   host->totalMem = host->usedMem = host->cachedMem = 0;
-   host->usedSwap = host->totalSwap = host->sharedMem = 0;
 
    pmAtomValue value;
-   if (Metric_values(PCP_MEM_TOTAL, &value, 1, PM_TYPE_U64) != NULL)
-      host->totalMem = value.ull;
    if (Metric_values(PCP_MEM_FREE, &value, 1, PM_TYPE_U64) != NULL)
       freeMem = value.ull;
    if (Metric_values(PCP_MEM_BUFFERS, &value, 1, PM_TYPE_U64) != NULL)
-      host->buffersMem = value.ull;
+      this->memValue[MEMORY_CLASS_BUFFERS] = value.ull;
    if (Metric_values(PCP_MEM_SRECLAIM, &value, 1, PM_TYPE_U64) != NULL)
       sreclaimableMem = value.ull;
    if (Metric_values(PCP_MEM_SHARED, &value, 1, PM_TYPE_U64) != NULL)
-      host->sharedMem = value.ull;
+      this->memValue[MEMORY_CLASS_SHARED] = value.ull;
    if (Metric_values(PCP_MEM_CACHED, &value, 1, PM_TYPE_U64) != NULL)
-      host->cachedMem = value.ull + sreclaimableMem - host->sharedMem;
-   const memory_t usedDiff = freeMem + host->cachedMem + sreclaimableMem + host->buffersMem;
-   host->usedMem = (host->totalMem >= usedDiff) ?
-           host->totalMem - usedDiff : host->totalMem - freeMem;
+      this->memValue[MEMORY_CLASS_CACHE] = value.ull + sreclaimableMem - this->memValue[MEMORY_CLASS_SHARED];
+   const memory_t usedDiff = freeMem + this->memValue[MEMORY_CLASS_CACHE] + sreclaimableMem + this->memValue[MEMORY_CLASS_BUFFERS];
+   this->memValue[MEMORY_CLASS_USED] = (super->totalMem >= usedDiff) ?
+           super->totalMem - usedDiff : super->totalMem - freeMem;
    if (Metric_values(PCP_MEM_AVAILABLE, &value, 1, PM_TYPE_U64) != NULL)
-      host->availableMem = MINIMUM(value.ull, host->totalMem);
+      this->memValue[MEMORY_CLASS_AVAILABLE] = MINIMUM(value.ull, super->totalMem);
    else
-      host->availableMem = freeMem;
+      this->memValue[MEMORY_CLASS_AVAILABLE] = freeMem;
    if (Metric_values(PCP_MEM_SWAPFREE, &value, 1, PM_TYPE_U64) != NULL)
       swapFreeMem = value.ull;
    if (Metric_values(PCP_MEM_SWAPTOTAL, &value, 1, PM_TYPE_U64) != NULL)
-      host->totalSwap = value.ull;
+      super->totalSwap = value.ull;
    if (Metric_values(PCP_MEM_SWAPCACHED, &value, 1, PM_TYPE_U64) != NULL)
-      host->cachedSwap = value.ull;
-   host->usedSwap = host->totalSwap - swapFreeMem - host->cachedSwap;
+      super->cachedSwap = value.ull;
+   super->usedSwap = super->totalSwap - swapFreeMem - super->cachedSwap;
+}
+
+static void PCPMachine_updateDarwinMemoryInfo(PCPMachine* this, Settings* settings) {
```

Comment:
Any way to reuse the `Platform_setMemoryValues` code in `darwin/Platform.c` for this?

I smell a WET (write everything twice). For every platform besides PCP, there now seems to be two copies of memory calculation formula, and I don't like this.